File: 3makingfriends/Kruskal.java

/* Minimal Spinning Tree */
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Scanner;
import java.util.Set;

public class Kruskal {

    List<Edge> edges;
    List<Set<Integer>> set;
    Node[] nodes;

    public static void main(String[] args) {
        new Kruskal().run();
    }

    public void run() {
        Parser p = new Parser(new Scanner(System.in));
        edges = p.edges;
        nodes = p.nodes;

        int result = kruskals();
        System.out.println(result);

    }

    private Node find(Node node) {
        Node p = node;
        while (p.parent != p) {
            p = p.parent;
        }
        while (node.parent != p) {
            Node w = node.parent;
            node.parent = p;
            node = w;
        }
        return p;
    }

    private int kruskals() {
        int total_weight = 0;
        Queue<Edge> queue = new PriorityQueue<>((e1, e2) -> e1.weight - e2.weight);
        queue.addAll(edges);
        while (!queue.isEmpty()) {
            Edge e = queue.peek();
            if (!find(e.u).equals(find(e.v))) {
                union(e);
                total_weight += e.weight;
            }
            queue.remove(e);
        }
        return total_weight;
    }

    private void union(Edge e) {
        Node u = find(e.u);
        Node v = find(e.v);
        if (u.size < v.size) {
            u.parent = v;
            v.size += u.size;
        } else {
            v.parent = u;
            u.size += v.size;
        }
    }

}

class Node {
    int id, size;
    Node parent;

    public Node(int id) {
        this.id = id;
        this.parent = this;
        this.size = 1;
    }
}

class Edge {

    Node u, v;
    int weight;

    public Edge(Node u, Node v, int weight) {
        this.u = u;
        this.v = v;
        this.weight = weight;
    }

    public String toString() {
        return u.id + "-" + v.id + "w:" + weight;
    }
}

class Parser {

    List<Edge> edges;
    Node[] nodes;

    public Parser(Scanner scan) {
        read(scan);
    }

    private void read(Scanner scan) {
        int numberNodes = scan.nextInt();
        int numberEdges = scan.nextInt();

        nodes = new Node[numberNodes];
        for (int i = 0; i < numberNodes; i++) {
            nodes[i] = new Node(i + 1);
        }

        edges = new ArrayList<>();
        for (int i = 0; i < numberEdges; i++) {
            int u = scan.nextInt();
            int v = scan.nextInt();
            int w = scan.nextInt();
            Edge edge = new Edge(nodes[u - 1], nodes[v - 1], w);
            edges.add(edge);
        }


        scan.close();
    }
}

File: 3makingfriends/Kruskal.py
from collections import deque
import heapq


class Edge:
    def __init__(self, u, v, weight):
        self.u = u
        self.v = v
        self.weight = weight

    def __lt__(self, other):
        return self.weight < other.weight

    def __eq__(self, other):
        return self.weight == other.weight

    def __hash__(self):
        return hash((self.u, self.v, self.weight))

    def __str__(self):
        return f"u:{self.u}, v:{self.v}, weight:{self.weight}"


class Kruskal:
    def __init__(self):
        self.edges = []
        self.nodes = []
        self.graph = {}
        self.parser()
        self.create_graph()
        self.kruskals()

    def parser(self):
        line = input().split()
        number_people = int(line[0])
        number_edges = int(line[1])
        for _ in range(number_edges):
            new_line = input().split()
            u = int(new_line[0])
            v = int(new_line[1])
            weight = int(new_line[2])
            edge = Edge(u, v, weight)
            self.edges.append(edge)
        for i in range(1, number_people + 1):
            self.nodes.append(i)

    def create_graph(self):
        for i in range(len(self.nodes)):
            self.graph[i] = []
        for edge in self.edges:
            self.graph[edge.u - 1].append(edge)
            self.graph[edge.v - 1].append(Edge(edge.v, edge.u, edge.weight))

    def kruskals(self):
        pred = {}
        distance = 0
        queue = []
        heapq.heapify(queue)

        for node, edges in self.graph.items():
            for e in edges:
                heapq.heappush(queue, e)

        while queue:
            e = queue[0]
            if self.find(pred, e.u) != self.find(pred, e.v):
                self.union(pred, e)
                distance += e.weight
            heapq.heappop(queue)
        print(distance)

    def find(self, pred, node):
        if node not in pred:
            return node
        if pred[node] != node:
            pred[node] = self.find(pred, pred[node])
        return pred[node]

    def union(self, pred, e):
        u = self.find(pred, e.u)
        v = self.find(pred, e.v)
        if u is not None and v is not None and u != v:
            pred[v] = u


if __name__ == "__main__":
    Kruskal()
